<a href="https://colab.research.google.com/github/rahiakela/deep_learning_for_vision_systems/blob/master/6-transfer-learning/1_pretrained_network_as_classifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Transfer learning approaches

There are three major transfer learning approaches as follows:

1. Pretrained network as a classifier
2. Pretrained network as feature extractor
3. Fine tuning

Each approach can be effective and save significant time in developing and training a deep convolutional neural network model. It may not be clear as to which usage of the pre-trained model may yield the best results on your new computer vision task, therefore some experimentation may be required.

## Setup

In [1]:
import tensorflow as tf
from tensorflow import keras
import numpy as np

from tensorflow.keras.preprocessing.image import load_img, img_to_array
from tensorflow.keras.applications.vgg16 import VGG16, preprocess_input, decode_predictions

import matplotlib.pyplot as plt
import matplotlib.cm as cm
%matplotlib inline

In [4]:
%%shell

wget -q https://github.com/rahiakela/deep_learning_for_vision_systems/raw/master/6-transfer-learning/dog.jpg
wget -q https://github.com/rahiakela/deep_learning_for_vision_systems/raw/master/6-transfer-learning/cat.jpg

## Pretrained network as a classifier

**The pre-trained model is used directly to classify new images with no changes applied to it and no extra training. All we do here is download the network architecture and its pretrained weights. Then run the predictions directly on our new data.** 

In this case, we are saying that the domain of our new problem is very similar to the one that the pretrained network was trained on and it is ready to just be “deployed”. So no training is done here.

In the dog breed example, we could have used VGG16 network that was trained on ImageNet dataset directly to run predictions. Because ImageNet already contains a lot dog breeds images so the significant portion of the representational power of the pretrained network may be devoted to features that are specific to differentiating between dog breeds.

**Using a pretrained network as a classifier doesn’t really involve any layers freezing or extra model training. Instead, it is just taking a network that was trained on your similar problem and deploying it directly to your task.**

Let’s see how to use a pretrained network as a classifier. In this example, we will use a VGG16 network that was pretrained on the ImageNet dataset to classify the image of the German Shepherd dog.

<img src='https://raw.githubusercontent.com/moelgendy/deep_learning_for_vision_systems/master/chapter_06/dog.jpg?raw=1' width='800'/>

Download the pretrained model of VGG16 and its ImageNet weights. We will set
include_top to True because we want to use the entire network as a classifier.

In [2]:
# load the VGG16 model
model = VGG16(weights='imagenet', include_top=True, input_shape=(224, 224, 3))

553467904/553467096 [==============================] - 4s 0us/step


Load and preprocess the input image

In [5]:
# load an image from file
image = load_img('dog.jpg', target_size=(224, 224))

# convert the image pixels to a numpy array
image = img_to_array(image)

# reshape data for the model
image = image.reshape((1, image.shape[0], image.shape[1], image.shape[2]))

# prepare the image for the VGG model
image = preprocess_input(image)

Now our input image is ready for us to run the predictions

In [6]:
# predict the probability across all output classes
yhat = model.predict(image)

# convert the probabilities to class labels
label = decode_predictions(yhat)

# retrieve the most likely result with the highest probability
label = label[0][0]

# print the classification
print('%s (%.2f%%)' % (label[1], label[2] * 100))

40960/35363 [==================================] - 0s 0us/step
German_shepherd (99.72%)


You can see that the model was already trained to predict the correct dog breed with a high confidence score (99.72%). This is because the ImageNet dataset has more than 20,000 labeled dog images and classified into 120 classes.

Let's test it with a cat image.

In [7]:
# load an image from file
image = load_img('cat.jpg', target_size=(224, 224))

# convert the image pixels to a numpy array
image = img_to_array(image)

# reshape data for the model
image = image.reshape((1, image.shape[0], image.shape[1], image.shape[2]))

# prepare the image for the VGG model
image = preprocess_input(image)

# predict the probability across all output classes
yhat = model.predict(image)

# convert the probabilities to class labels
label = decode_predictions(yhat)

# retrieve the most likely result with the highest probability
label = label[0][0]

# print the classification
print('%s (%.2f%%)' % (label[1], label[2] * 100))

Egyptian_cat (65.18%)
